### PR TITLE
Refactoring of TF based augmentations

### PR DESCRIPTION
**Pull request by tilmankamp**

### CODE REVIEW

File: doc/TRAINING.rst [hunk 1]
```diff
@@ -297,14 +291,31 @@ In the documentation below, whenever a value is specified as ``<float-range>`` o
...
+.. warning::
+    If feature caching is enabled and infinite (default), these augmentations will only be performed on first epoch and the result will be reused for subsequent epochs. This would not only hinder value ranges from reaching their intended final values, but could also lead to unintended over-fitting. In this case flag ``--cache_for_epochs N`` (with N > 1) should be used to periodically invalidate the cache and thus allow samples to be re-augmented in new ways and with current range-values.
+
+Every augmentation is targeting a certain data representation of the sample - further on called *domain*.
+Augmentations are applied domain-wise in the following order:
+
+1. **sample** domain: The sample just got loaded and its waveform is represented as a NumPy array. For implementation reasons these augmentations are the only ones that can be "simulated" through ``bin/play.py``.
+
+2. **signal** domain: The sample waveform is represented as a tensor.
```
[the remark this answers is not in the document]
Comment (tilmankamp) [reply doc/TRAINING.rst:308]: The only ones are "multiple".

File: doc/TRAINING.rst [hunk 2]
```diff
@@ -361,25 +362,123 @@ If feature caching is enabled, these augmentations will only be performed on the
...
+
+  * **n**: number of intervals to mask
+
+  * **size**: number of frequency bands to mask per interval
+
+Multi domain augmentations
+--------------------------
+
+**Time mask augmentation** ``--augment time_mask[p=<float>,n=<int-range>,size=<float-range>,domain=<domain>]``
+  Sets time-intervals within the augmented samples to zero (silence) at random positions.
```
[the remark this answers is not in the document]
Comment (tilmankamp) [reply doc/TRAINING.rst:415]: Thanks for the link!

File: training/deepspeech_training/util/helpers.py [hunk 3]
```diff
@@ -174,3 +174,18 @@ def pick_value_from_range(value_range, clock=None):
     value = value_range.start + clock * (value_range.end - value_range.start)
     value = random.uniform(value - value_range.r, value + value_range.r)
     return round(value) if isinstance(value_range.start, int) else value
+
+
+def tf_pick_value_from_range(value_range, clock=None, double_precision=False):
+    import tensorflow as tf  # pylint: disable=import-outside-toplevel
```
[the remark this answers is not in the document]
Comment (tilmankamp) [reply training/deepspeech_training/util/helpers.py:180]: Importing TF for the first time comes with huge overhead and a tremendous amount of logging to the console. This would seriously affect UX for `bin/play.py`. The alternative approach of having augmentations that require TF in a separate file would produce unnecessarily complicated code. That code would also have to rely on late binding of the additional module. I came to the conclusion that having a handful of local imports is more KISS.

[the remark this answers is not in the document]
Comment (reuben) [reply training/deepspeech_training/util/helpers.py:180]: > Importing TF for the first time comes with huge overhead and a tremendous amount of logging to the console. This would seriously affect UX for bin/play.py.

play.py could also override the TF logging config like train.py does (which reminds me, we should port that to evaluate.py as well).

> That code would also have to rely on late binding of the additional module. I came to the conclusion that having a handful of local imports is more KISS.

Agreed on simpler solution. Can you add a brief comment just saying something like "Late import of TF to avoid logging spam during program initialization".

File: training/deepspeech_training/util/augmentations.py [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+        super(GraphAugmentation, self).__init__(p)
+        if domain not in ['signal', 'spectrogram', 'features']:
+            raise ValueError('Unsupported augmentation domain: {}'.format(domain))
+        self.domain = domain
+
+    def apply(self, tensor, clock=0.0):
+        raise NotImplementedError
+
+    def apply_with_probability(self, tensor, clock=0.0):
+        import tensorflow as tf  # pylint: disable=import-outside-toplevel
```
[the remark this answers is not in the document]
Comment (tilmankamp) [reply training/deepspeech_training/util/augmentations.py:43]: Same as explained below.

File: training/deepspeech_training/util/augmentations.py [hunk 5]
```diff
@@ -0,0 +1,564 @@
...
+
+def _enqueue_overlay_samples(sample_source, queue, buffering=BUFFER_SIZE):
+    """
+    As the central distribution point for overlay samples this function is supposed to run in one process only.
+    This ensures that samples are not used twice if not required.
+    It loads the (raw and still compressed) data and provides it to the actual augmentation workers.
+    These are then doing decompression, potential conversion and overlaying in parallel.
+    """
+    # preventing cyclic import problems
+    from .sample_collections import samples_from_source  # pylint: disable=import-outside-toplevel
```
[the remark this answers is not in the document]
Comment (tilmankamp) [reply training/deepspeech_training/util/augmentations.py:237]: Nice catch!

File: doc/TRAINING.rst [hunk 6]
```diff
@@ -361,25 +362,104 @@ If feature caching is enabled, these augmentations will only be performed on the
...
         python -u DeepSpeech.py \
           --train_files "train.sdb" \
-          --augmentations_per_epoch 10 \
+          --feature_cache ./feature.cache \
+          --cache_for_epochs 10 \
+          --epochs 100 \
           --augment overlay[p=0.5,source=noise.sdb,layers=1,snr=50:20~10] \
-          --augment overlay[p=0.2,source=voices.sdb,layers=10:6,snr=50:20~10] \
           --augment reverb[p=0.1,delay=50.0~30.0,decay=10.0:2.0~1.0] \
-          --augment gaps[p=0.05,n=1:3~2,size=10:100] \
```
Comment (DanBmh) [inline doc/TRAINING.rst:375]: Why did you remove the gaps augmentation? Is it similar to time_mask?

File: doc/TRAINING.rst [hunk 7]
```diff
@@ -361,25 +362,104 @@ If feature caching is enabled, these augmentations will only be performed on the
...
 .. code-block:: bash
 
         python -u DeepSpeech.py \
           --train_files "train.sdb" \
-          --augmentations_per_epoch 10 \
+          --feature_cache ./feature.cache \
+          --cache_for_epochs 10 \
+          --epochs 100 \
           --augment overlay[p=0.5,source=noise.sdb,layers=1,snr=50:20~10] \
-          --augment overlay[p=0.2,source=voices.sdb,layers=10:6,snr=50:20~10] \
```
Comment (DanBmh) [inline doc/TRAINING.rst:373]: I think its a good idea to show that you can use the flag multiple times here, for noise and speech with different params.

File: doc/TRAINING.rst (hunk 6 above)
Comment (tilmankamp) [reply doc/TRAINING.rst:375]: Yes
